internal/nomad/job.go

```diff
@@ -99,7 +99,11 @@ func FindOrCreateConfigTask(taskGroup *nomadApi.TaskGroup) *nomadApi.Task {
 	if task.Config == nil {
 		task.Config = make(map[string]interface{})
 	}
-	task.Config["command"] = ConfigTaskCommand
+	// This function should allow concurrency in the "Find" case.
+	// Therefore, this condition is necessary to remove concurrent writes in the "Find" case.
+	if v, ok := task.Config["command"]; !(ok && v == ConfigTaskCommand) {
+		task.Config["command"] = ConfigTaskCommand
+	}
```

Comment:
I am not sure whether I understand these changes. I get that you want to enable concurrency when an existing config job is found and does need to be created. And you want to prevent concurrent writes to the command config of the task.

Now, I am wondering if the code is able to achieve that: What if thread A checks the condition and enters the `if branch` just before line 105 gets executed, another thread B also checks the condition and jumps into line 105. In that case, both threads might still execute line 105, possibly with a concurrent write. Similarly, this could affect the `if task == nil` and `if task.Config == nil` checks. My search for occurrences of concurrent access to these lines was not super successful (without checking the tests but including other callees). Could you please help me out here (and for the similar lines below)?

Reply:
> What if ...

You are absolutely right!, except we never enter one of the conditions.
These conditions are just guarding clauses but since we have a correct `template-environment-job.hcl` (that includes a config and a default task) and we only edit jobs the correct way, we never use them.

> My search for occurrences of concurrent access to these lines was not super successful

Whenever new runners are requested, we make a copy of the template job. In this process we enter the config (and default) task. In a concurrent way this can happen when two users request a runner in the exact same moment. Noticeable by a slightly flaky behaviour of the `TestTwoSampleAddExactlyTwoRunners` test.

Maybe I shouldn't have smuggled this into this merge request and maybe it would make more sense to validate the job when creating an environment than to do this every time it is accessed.

Reply:
Okay, I think we're discussing two cases here: The first is about the usage of the guard clause and the second about the concurrent access to the `FindOrCreateConfigTask`. I'll try to address them one by another:

1. Guard Clause
    Okay. The idea is that we only change the value of `task.Config["command"]` if it is not our expecting value of `ConfigTaskCommand`, is it?

2. Concurrent Access
    If I get you right, you argue that the concurrent access might happen if two users request a new runner at the same time (which yes, is a valid use case). In this case, I am wondering if the concurrent write is actually possible: I thought that each runner gets cloned first and any changes are only made to the copy of the template job. This would imply that two threads cannot modify the same object at the same time (because each runner request would first create a copy [read-only access to the template] and then modify the dedicated copy of the template in memory [if required]). And the environment should not be affected by this change, is it?

Probably, the question regarding the concurrent access is the difficult question for me. I am sorry for nagging you here and really tried to have a look on my own but still was unable to find answers to these questions.

Reply:
Thanks for your comments (and the renaming), they persist the understanding I was able to gather throughout this PR and the discussion we had🦉.

Reply:
1. Yes
2. The copy was not read only as the getters access the tasks and those access functions (`FindOrCreateConfigTask`) were always overwriting the task config.

As discussed I added some documentation.